### ISSUE TITLE
Missing annotation in function passed in a click decorator

### ISSUE THREAD

**Issue by diamond-lizard** [python/mypy#3869]
When I run _"mypy --disallow-any unannotated [foo.py](https://gist.github.com/diamond-lizard/033afc80c97b7cb84913176a259f2b37)",_ mypy tells me that _"[foo.py](https://gist.github.com/diamond-lizard/033afc80c97b7cb84913176a259f2b37):6: error: Function is missing a return type annotation"_.

[Here](https://gist.github.com/diamond-lizard/1bb89d6d80b387e1cfc452aa797f3814) is the output of running _"mypy -v --disallow-any unannotated [foo.py](https://gist.github.com/diamond-lizard/033afc80c97b7cb84913176a259f2b37)"_

I'm not sure if this is a bug or if there's some special way I'm supposed to annotate the line mypy is complaining about.

**Comment by ethanhs** [comment]
The issue is `cli` is supposed to be annotated with a return type. This appears to be an issue with the line number of the reported error.

**Comment by diamond-lizard** [comment]
Thank you for your help.

When I annotated the `cli` function itself with `a -> None` annotation, the error disappeared.  The issue does seem to be with the reported line number, as you say.  If the line number had been reported correctly, the necessary annotation would have been obvious.

**Comment by ethanhs** [comment]
I believe we just increment the line number of the last decorator to find the function line number, which is not always accurate. We also cannot decrement from the first statement, as there may be a blank line between the nodes. Looking at the raw information of the ast parsing, I don't believe there is information on the line the annotation _should_ be if it is absent (since the attribute `returns` is `None`). This might require a change to typed_ast.

**Comment by gvanrossum** [comment]
Surely there's a line number in the AST for the 'def'? That should be less
confusing.

**Comment by ethanhs** [comment]
The `FunctionDef` node has a line number of 4, the line number of the first decorator. The line number of the first `def` token is 10. However this is also not the logical place, consider:

```python
def foo(a: int,
             b: str): ...
```

The returns should be on line 2, however the `lineno` of the `FunctionDef` would be line 1.

**Comment by diamond-lizard** [comment]
When mypy informs me that a _"Function is missing a return type annotation"_, all I'm wondering is "which function?"  Telling me which line the def is on answers that question.  From there, I can figure out on my own that the type annotation itself should go after the parenthesis that closes the parameter list, whichever line that might be on.

If mypy pointed to the exact place where the annotation should go it would indeed be a nice bit of sugar, but the essential information seems to be which line the function in question is on.

What confused me about the error pointing to line 6 in the original [foo.py](https://gist.github.com/diamond-lizard/033afc80c97b7cb84913176a259f2b37) program was that it pointed a place in the decorator that contained a function call, so I (and another person helping me out on #python) both wondered if mypy was complaining that it was the function called on line 6 -- click.File() -- that mypy thought was lacked a type annotation.  If the error just pointed to the def, there would be no such confusion.

**Comment by ethanhs** [comment]
Yes, I understand the issue, I am unsure of how to solve it internally in mypy. We use the typed_ast module to get the line number information. I don't think the AST exposes the location of the return type, which we want.

Also, in future, if you want help on mypy, you may be interested in the chat https://gitter.com/python/typing . (If you prefer IRC, there is an IRC bridge for gitter). There is almost always someone around who can help out with this.

**Comment by gvanrossum** [comment]
Dang. I didn't remember the Python AST doesn't have a separate node for the
'def' (it exists in the grammar but it's optimized out in the AST). Can you
file a new mypy issue and one in typed_ast?

**Comment by hauntsaninja** [comment]
A lot has changed since 2017, e.g. see #6753
